lib/spack/spack/directives.py

```diff
@@ -758,6 +760,17 @@ def _execute_resource(pkg):
     return _execute_resource
 
 
+def buildsystem(*values, **kwargs):
+    default = kwargs.get("default", None) or values[0]
+    return variant(
```

Comment:
I still don't like the idea of modeling this as a variant. For 99.99% of packages, the package will either use a single build system, or the build system will be dependent on package version or OS. For the other 0.01% of packages, do we really _need_ to support user-configurable build systems? It seems like a lot of added complexity that isn't needed for most packages. My design philosophy is to make 90% of things easy, 9% possible, and 1% impossible. This feels like it falls under the 1%.

Reply:
> I still don't like the idea of modeling this as a variant. [ ... ] This feels like it falls under the 1%.

I think this is also a misunderstanding on the current architecture of the PR. This design uses a variant because:

1. A build-system fits the semantic of a variant and a variant allows users to refer to it in a spec literal
2. We can make dependencies conditional on the build-system (which is what we do in base classes)

For 1. there are use cases like `arpack-ng` that maintains two different build-systems unconditionally. Using a variant we can discriminate between `arpack-ng buildsystem=autotools` and `arpack-ng buildsystem=cmake`.[^1] 

Point 2. instead allows us to define base classes like:

https://github.com/spack/spack/blob/3b87584e6194e02bd4c0e75d3b8c181ab9980dea/lib/spack/spack/build_systems/cmake.py#L54-L72

where build-system specific dependencies are conditional on the variant. This ensures that when we have multiple build-systems we won't pull-in dependencies we don't need from other build-systems (which is what happens right now).

Also **a package that has a single build-system has the directive and dependencies defined in the base `*Package` class**, so there's no boilerplate to be cut and pasted over and over. The only difference in the long term is the requirement to code the installation procedure in a separate class (which would allow us to delete a few hundred lines related to the run-time generated adapter).

[^1]: It's possible we might see cases like that also happen in bootstrapping contexts (`foo` can be built with CMake and XXX, CMake depends on `foo buildsystem=XXX` to avoid a dependency on itself).

Reply:
> there are use cases like `arpack-ng` that maintains two different build-systems unconditionally.

Yes, but I feel like this use case falls under the 1%. I guess my question is, do we really _need_ to support this use case? Are there some users who want CMake and other users who want Autotools? In cases like Windows where we have to use CMake, that's fine, but do we really need to allow users to choose their own build system?

I can understand having a `build_system` variant for these 1% of cases if users are desperate for this feature, but do we also need to have a `build_system` variant for the other 99% of cases too? In my mind, a variant is an option. If there's only one option for 99% of packages, that doesn't feel like it should be a variant.

> ```python
> buildsystem("cmake") 
> with when("buildsystem=cmake"): 
> ```

This looks odd to me. But I'm guessing it's because `buildsystem` can be overridden by a subclass. I guess `buildsystems` aren't additive like `depends_on`?

From reading the rest of your comments, I'm guessing that the reason for using a variant is more technical than anything else. That is, there isn't really a better way to model all of the things we need (check the current build system from the spec). And since we can't access `self` in `depends_on`, we have to expose the build system via a variant. Kind of an unfortunate limitation of our current API, but not something we can easily change. So I guess I'm fine with this since there isn't really an alternative, but I just wanted to express my impression that this felt like an odd use for a variant.

Reply:
> I just wanted to express my impression that this felt like an odd use for a variant.

I agree with treating as a variant because users who inherit one build system will never have to think of it as a variant. If users ever need to check this in multi-buildsystem cases, I think the variant will make more sense (it has implications for how the package is built, so would be another detail to check about the spec).